stub/src/main/java/io/grpc/stub/BlockingClientCall.java

```diff
@@ -0,0 +1,407 @@
+      throw new RuntimeException(e); // should never happen
+    }
+  }
+
+  /**
+   * Send a value to the stream for sending to server, wait if necessary for the grpc stream to be
+   * ready up to specified timeout.
+   * <br>
+   * If write is not legal at the time of call, immediately returns false
+   * <br><br>
+   * <p>
+   * <b>NOTE:  </b>This method will return as soon as it passes the request to the grpc stream
+   * layer. It will not block while the message is being sent on the wire and returning true does
+   * not guarantee that the server gets the message.
+   * </p>
+   * <p>
+   * <b>WARNING:  </b>Doing only writes without reads can lead to deadlocks as a result of flow
+   * control.  Furthermore, the server closing the stream will only be identified after the last
+   * sent value is read.
+   * </p>
+   *
+   * @param request Message to send to the server
+   * @param timeout How long to wait before giving up.  Values &lt;= 0 are no wait
+   * @param unit A TimeUnit determining how to interpret the timeout parameter
+   * @return true if the request is sent to stream, false if skipped
+   * @throws TimeoutException if write does not become ready before the specified timeout expires
+   * @throws io.grpc.StatusRuntimeException If the stream has closed in an error state
+   */
+  public boolean write(ReqT request, long timeout, TimeUnit unit)
+      throws InterruptedException, TimeoutException {
+
+    if (writeClosed) {
+      throw new IllegalStateException("Writes cannot be done after calling halfClose or cancel");
+    }
+
+    boolean writeDone = false;
+    long start = System.nanoTime();
+    long duration = unit.toNanos(timeout);
+
+    while (!writeDone && isWriteLegal()) {
+      if (call.isReady()) {
+        call.sendMessage(request);
+        writeDone = true;
+      } else {
+        waitAndDrainExecutorOrTimeout(start, duration);
+      }
+    }
+
+    if (writeDone || getClosedStatus() == null) {
+      return writeDone;
+    }
+
+    if (getClosedStatus().isOk()) {
+      return false;
+    } else {
+      // Propagate any errors returned from the server
+      throw getClosedStatus().asRuntimeException();
+    }
+  }
+
+  /**
+   * Cancel stream and stop any further writes.  Note that some reads that are in flight may still
+   * happen after the cancel.
+   *
+   * @param message if not {@code null}, will appear as the description of the CANCELLED status
+   * @param cause if not {@code null}, will appear as the cause of the CANCELLED status
+   */
+  public void cancel(String message, Throwable cause) {
+    if (message == null) {
+      message = "User requested a cancel";
+    }
+    call.cancel(message, cause);
+    writeClosed = true;
+    executor.add(NoOpRunnable.INSTANCE);
+  }
+
+  /**
+   * Indicate that no more writes will be done and the stream will be closed from the client side.
+   * <br>
+   * See {@link ClientCall#halfClose()}
+   */
+  public void halfClose() {
```

Comment:
I am wondering, from API perspective, if the call is halfclosed, but buffer still has some value, should a read return anything, or we clean things up?